File: 130-surrounded-regions/solution.cpp
#include <vector>

class Solution
{
public:
  void solve(std::vector<std::vector<char>> &board)
  {
    const int row_len = board.size(), col_len = board[0].size();
    std::vector<std::pair<int, int>> to_check;
    for (int row = 0; row < row_len; ++row)
    {
      to_check.emplace_back(row, 0);
      to_check.emplace_back(row, col_len - 1);
    }

    for (int col = 0; col < col_len; ++col)
    {
      to_check.emplace_back(0, col);
      to_check.emplace_back(row_len - 1, col);
    }

    while (!to_check.empty())
    {
      const auto row_col = to_check.back();
      const auto row = row_col.first;
      const auto col = row_col.second;
      to_check.pop_back();
      if (row < 0 || row >= row_len || col < 0 || col >= col_len || board[row][col] != 'O')
      {
        continue;
      }

      board[row][col] = 'U';
      to_check.emplace_back(row - 1, col);
      to_check.emplace_back(row + 1, col);
      to_check.emplace_back(row, col - 1);
      to_check.emplace_back(row, col + 1);
    }

    for (int row = 0; row < row_len; ++row)
    {
      for (int col = 0; col < col_len; ++col)
      {
        if (board[row][col] == 'U')
        {
          board[row][col] = 'O';
        }
        else if (board[row][col] == 'O')
        {
          board[row][col] = 'X';
        }
      }
    }
  }
};

File: 130-surrounded-regions/solution.py
class Solution:
  def solve(self, board: list[list[str]]) -> None:
    """
    Do not return anything, modify board in-place instead.
    """

    row_len, col_len = len(board), len(board[0])
    to_check: list[tuple[int, int]] = []
    for row in range(row_len):
      to_check.extend([(row, 0), (row, col_len - 1)])

    for col in range(col_len):
      to_check.extend([(0, col), (row_len - 1, col)])

    while to_check:
      row, col = to_check.pop()
      if row < 0 or row >= row_len or col < 0 or col >= col_len or board[row][col] != 'O':
        continue

      board[row][col] = 'U'  # unchangeable
      to_check.extend([(row - 1, col), (row + 1, col),
                      (row, col - 1), (row, col + 1)])

    for row in range(row_len):
      for col in range(col_len):
        if board[row][col] == 'U':
          board[row][col] = 'O'
        elif board[row][col] == 'O':
          board[row][col] = 'X'
